parsons/ngpvan/email.py

```diff
@@ -0,0 +1,143 @@
+
+logger = logging.getLogger(__name__)
+
+
+class Email(object):
+    def __init__(self, van_connection):
+
+        self.connection = van_connection
+
+    def get_emails(
+        self,
+        orderby=None,
+    ):
+        """
+        Get emails.
+
+        Note that it takes some time for the system to aggregate opens and click-throughs,
+        so data can be delayed up to 15 minutes.
+
+        `Args:`
+            orderby : str
+                Optional; sorts results in ascending (asc) or descending (desc) order
+                for the dateModified field.
+
+        `Returns:`
+            Parsons Table
+                See :ref:`parsons-table` for output options.
+        """
+
+        params = {
+            "$orderby": orderby,
+        }
+
+        tbl = Table(self.connection.get_request("email/messages", params=params))
+        logger.debug(f"Found {tbl.num_rows} emails.")
+        return tbl
+
+    def get_email(self, email_id, expand=True):
+        """
+        Get an email.
+
+        `Args:`
+            email_id : int
+                The email id.
+            expand : bool
+                Optional; expands the email message to include the email content and
+                statistics. Defaults to True.
+
+        `Returns:`
+            dict
+        """
+
+        params = {
+            "$expand": (
+                "emailMessageContent, EmailMessageContentDistributions"
+                if expand
+                else None
+            ),
+        }
+
+        r = self.connection.get_request(f"email/message/{email_id}", params=params)
+        logger.debug(f"Found email {email_id}.")
+        return r
+
+    def get_email_stats(self):
+        """
+        Get stats for all emails.
+
+        `Args:`
+            emails : list
+                A list of email message details.
+
+        `Returns:`
+            Parsons Table
+                See :ref:`parsons-table` for output options.
+        """
+
+        email_list = []
+
+        final_email_list = []
+
+        emails = self.get_emails()
+
+        foreign_message_ids = [email["foreignMessageId"] for email in emails]
+
+        for fmid in foreign_message_ids:
+            email = self.get_email(fmid)
+            email_list.append(email)
+
+        for email in email_list:
+            d = {}
+            d["name"] = email["name"]
+            d["createdBy"] = email["createdBy"]
+            d["dateCreated"] = email["dateCreated"]
+            d["dateModified"] = email["dateModified"]
+            d["dateScheduled"] = email["dateScheduled"]
+            d["foreignMessageId"] = email["foreignMessageId"]
+            d["recipientCount"] = 0
+            d["bounceCount"] = 0
+            d["contributionCount"] = 0
+            d["contributionTotal"] = 0
+            d["formSubmissionCount"] = 0
+            d["linksClickedCount"] = 0
+            d["machineOpenCount"] = 0
+            d["openCount"] = 0
+            d["unsubscribeCount"] = 0
+            try:
```

Comment:
1. Why do we need a try block here? What issues have we run into?
2. If we need it, would it maybe be better to have each of these assignments get their own try block so that - if something breaks with `bounceCount` (for example) - it doesn't affect all of the other assignment commands?

Reply:
I believe the problems I was running into was that when emails weren't sent yet, these fields did not exist. This was also code I wrote pretty early on in my career and a few months ago added it to parsons without looking too close into it because I was about to lose access to EveryAction and wanted to contribute before that went away.

Reply:
Got it. Do we expect all of these fields to be missing simultaneously in the same `i` block?

Another question - are there cases where we would expect to loop multiple times?  Does this happen for emails that have A/B tests?

Reply:
I believe so. 

Yes, if there are A/B tests it loops multiple times.

Reply:
Okay, then the try block should be fine.

Would it be possible to have the option of not aggregating over each email? For the reporting we do for Tech for Campaigns, we report on how each send did independently.

Reply:
Yeah, we could create another method or could have a param that gives the option.

Reply:
As per our discussion on 2024-05-14 we can defer this request to a later PR.